### PR TITLE
elf: fast track when the host used matches the base

### PR DESCRIPTION
**Pull request by sergiusens**
LP: #1746607

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

https://bugs.launchpad.net/snapcraft/+bug/1746607

### CODE REVIEW

**Comment by kyrofa** [comment]
I haven't fully reviewed yet, but here's a quick pass on numbers priming the `ros-talker-listener` integration test (just priming, nothing else):

This PR (snapped):
real	5m52.839s
user	2m23.464s
sys	3m18.844s

Stable (2.39):
real	5m58.006s
user	2m24.372s
sys	3m21.072s

2.35:
real	1m39.451s
user	1m17.740s
sys	0m9.492s

I sort of expected this one to get us back to 2.35 levels, but that doesn't seem to be the case.

File: snapcraft/_options.py [hunk 1]
```diff
@@ -223,6 +225,16 @@ def __init__(self, use_geoip=False, parallel_builds=True,
         self._set_machine(target_deb_arch)
         self.__debug = debug
 
+    @property
+    def is_host_same_base(self):
+        codename = None
+        with suppress(errors.OsReleaseCodenameError):
+            codename = OsRelease().version_codename()
+
+        logger.debug('Running on {!r}'.format(codename))
```
Comment (elopio) [inline snapcraft/_options.py]: A nit, but this debug message will look ugly when codename is None.

**Comment by elopio** [comment]
The autopkgtests comparisson is not looking good: https://paste.ubuntu.com/p/wt8cb9X5Ms/
There might be a lot of reasons for these tests to be slow, but this tells us to be cautious.
I see two nice improvements in this branch, so the change looks good to me. But to be sure about the savings we would need a more controlled experiment.

**Comment by sergiusens** [comment]
That new error is irrelevant, it is a timeout on a go project.

**Comment by sergiusens** [comment]
@elopio yeah, I don't trust those numbers as much

**Comment by elopio** [comment]
These are my numbers with the ros-talker-listener: https://paste.ubuntu.com/p/c2d4WrMKfz/

**Comment by kyrofa** [comment]
@elopio those numbers are pretty different than mine. Is that just the prime step?

**Comment by elopio** [comment]
Yes, on my canonistack dev machine.

**Comment by sergiusens** [comment]
master/512c54488e148b4b9778f2fe62640f6eee4b2a2b
```
real    6m0.049s
user    3m21.753s
sys     1m56.164s
```

```
real    1m40.039s
user    1m21.506s
sys     0m18.502s
```

Closing this in favour of smaller changes, starting with #1925 for 2.39.1, the more aggressive ones for later.

**Pull request closed without merging**